### ISSUE TITLE
createBreakpoints causes incompatibilities with ThemeOverride type

### ISSUE THREAD

**Issue by kristian240** [chakra-ui/chakra-ui#3372]
# 🐛 Bug report

After I updated to `@chakra-ui/react@1.3.2` I kept getting this error 

```
Type '{ styles: { global: { [x: string]: { height: string; width: string; fontFamily: string; bg: string; color: string; overflowX: string; scrollBehavior?: undefined; verticalAlign?: undefined; boxSizing?: undefined; } | { ...; } | { ...; } | { ...; }; body: { ...; }; svg: { ...; }; '*': { ...; }; }; }; fonts: { ...; }; ...' is not assignable to type 'ThemeOverride'.
  Type '{ styles: { global: { [x: string]: { height: string; width: string; fontFamily: string; bg: string; color: string; overflowX: string; scrollBehavior?: undefined; verticalAlign?: undefined; boxSizing?: undefined; } | { ...; } | { ...; } | { ...; }; body: { ...; }; svg: { ...; }; '*': { ...; }; }; }; fonts: { ...; }; ...' is not assignable to type 'DeepThemeExtension<{ components: { Accordion: { parts: string[]; baseStyle: { container: { borderTopWidth: string; borderColor: string; _last: { borderBottomWidth: string; }; }; button: { fontSize: string; _focus: { boxShadow: string; }; _hover: { ...; }; _disabled: { ...; }; px: number; py: number; }; panel: { ...;...'.
    The types of 'breakpoints.toString' are incompatible between these types.
      Type '(() => string) & (() => string)' is not assignable to type 'string'.
```

and my breakpoints look like this

```
import { createBreakpoints } from '@chakra-ui/theme-tools';

const breakpoints = createBreakpoints({
  sm: '375px',
  md: '580px',
  lg: '768px',
  xl: '1024px',
  '2xl': '1280px',
});

export default breakpoints;
```

I tried multiple things to fix this issue but no luck. 

Things I tried: 
* chakra-cli -> no luck since the code couldn't be executed since this is a TS error
* Using breakpoints from chakra's default theme but again same error occurred
* Different Typescript versions

## 💥 Steps to reproduce

You cannot build the application at all 🙈 

## 💻 Link to reproduction

I was not able to reproduce the error in codesandbox and the repo where the error has occurred is a private one.

## 🧐 Expected behavior

> Tell us what you expect to happen.

## 🧭 Possible Solution

> Not compulsory, but suggesting a fix would be great!

## 🌍 System information

| Software         | Version(s) |
| ---------------- | ---------- |
| Chakra UI        |        1.3.2    |
| Browser          |       -     |
| Operating System |      MacOS BigSur 11.2.1      |

I am using:
```
"@chakra-ui/react": "^1.3.2",
"@emotion/react": "^11.1.5",
"@emotion/styled": "^11.1.5",
"next": "^10.0.5",
"react": "^17.0.1",
"react-dom": "^17.0.1",
"typescript": "^4.1.3"
```

## 📝 Additional information

If I downgrade the `chakra-ui/react` I don't get the error anymore.

**Comment by 3runoDesign** [comment]
this in reality when using the

```ts


const theme = extendTheme({breakpoints});
```

**Comment by TimKolberger** [comment]
Hi @kristian240

This issue was fixed with @chakra-ui/react@1.3.2.
Could you check with the new version and close the issue if it is fixed for you as well?

**Comment by YuhriBernardes** [comment]
I'm having the same problem.

The error returned from it is:

```
Type 'Breakpoints<{ base: string; sm: string; md: string; lg: string; xl: string; '2xl': string; }>' is not assignable to type 'string[] & Record<string, any> & { base: "0em"; } & DeepThemeExtension<Partial<Breakpoints<{ sm: string; md: string; lg: string; xl: string; "2xl": string; }>>, Breakpoints<Record<string, any>>>'.
  Type 'Breakpoints<{ base: string; sm: string; md: string; lg: string; xl: string; '2xl': string; }>' is not assignable to type 'DeepThemeExtension<Partial<Breakpoints<{ sm: string; md: string; lg: string; xl: string; "2xl": string; }>>, Breakpoints<Record<string, any>>>'.
    Types of property 'toString' are incompatible.
      Type '(() => string) & (() => string)' is not assignable to type 'string'.ts(2322)
theme.types.d.ts(63, 5): The expected type comes from property 'breakpoints' which is declared here on type 'ThemeOverride'
```

There is the code:

```js
import { extendTheme } from '@chakra-ui/react'
import { createBreakpoints } from '@chakra-ui/theme-tools'

const breakpoints = createBreakpoints({
  base: '0em',
  sm: '960px',
  md: '1024px',
  lg: '1440px',
  xl: '1960px',
  '2xl': '2600px'
})

export const theme = extendTheme({
  breakpoints,
  colors: {
    black: '#16171B',
    blackAlpha: {
      900: '#202125'
    }
  },
  fonts: {
    body: 'Roboto, sans-serif',
    heading: 'Roboto, sans-serif'
  }
})
```

**Comment by TimKolberger** [comment]
I tried it with the following `package.json` and both provided custom themes in this issue.
I did not experience any error while running dev (CRA), build, chakra-cli nor `tsc --noEmit` 🤔 

Could you kindly recheck your installed versions of `@chakra-ui/react` and `@chakra-ui/cli` (if used)?
If you are still experiencing issues, it would be a great help if you provide a reproduction repo for this. 

```json
{
  "dependencies": {
    "@emotion/react": "^11.1.5",
    "@emotion/styled": "^11.1.5",
    "@types/node": "^14.14.27",
    "@types/react": "^17.0.2",
    "@types/react-dom": "^17.0.1",
    "framer-motion": "^3.3.0",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-icons": "^3.0.0",
    "react-scripts": "4.0.2",
    
    "@chakra-ui/react": "^1.3.3",
    "typescript": "^4.1.5"
  },
  "devDependencies": {
    "@chakra-ui/cli": "^1.2.1"
  },
  "scripts": {
    "gen:theme-typings": "chakra-cli tokens src/theme.ts",
    "postinstall": "npm run gen:theme-typings"
  }
}
```

**Comment by kristian240** [comment]
*TL;DR;* Use `strict: true` in your `tsconfig` file.

Hi @TimKolberger. I updated the issue and that didn't seemed to work. However I was trying to reproduce the issue on CodeSandbox (there is no issue by default) and noticed that mine tsconfig and default one there aren't the same. So I played with it a bit and found out that this issue occurs when `strict: false` is being used. 

So with this said, I created a sandbox and here it is https://codesandbox.io/s/sleepy-sanne-t3fk4?file=/src/App.tsx

**Comment by kristian240** [comment]
@TimKolberger I played a bit with this and it seems that you can instead of changing `strict` from `false` to `true` you can add this `strictNullChecks: true`. 

I hope this is helpful and that this will be fixed in future, because I don't see why is there an error with `false` and not with `true`. If anything it would make sence to be other way around 😄

**Comment by TimKolberger** [comment]
Thank you for your investigation @kristian240! This is surely something that should be fixed, and I am glad that you discovered a workaround in the meantime. 

Kind of odd that it only happens with `strictNullCheck: false` 😄

**Comment by AdamGilliesGit** [comment]
Hey, fairly new to Chakra-ui but I changed:

**Comment by AdamGilliesGit** [comment]
Hey, fairly new to Chakra-ui but changing the following code in breakpoint.d.ts fixed the type issue for me.

`export declare type Breakpoints<T = BaseBreakpointConfig> = string[] & WithBase<T>;`

To:

`export declare type Breakpoints<T = BaseBreakpointConfig> = Record<string, string> & WithBase<T>;`

Not much use if you are using build pipelines though.

**Comment by hutber** [comment]
I'm with Tim on this one :D Thank you very much for checking though @kristian240 I am going to see now if that fixes our problem.
But this does feel like a "breaking change" in that... its broken our build and we cannot upgrade a patch :p

**Comment by kristian240** [comment]
Good luck @hutber If I can help anything let me know 🤗

**Comment by hutber** [comment]
@TimKolberger Hey Tim, I just thought I'd reach out and see if you know if the team has any kind of roadmap for fixing this?

**Comment by hutber** [comment]
So I can get the error coming through on my IDE which was not happening before:

import { extendTheme, ThemeOverride } from "@chakra-ui/react"
const appTheme: ThemeOverride = {
  styles,
  breakpoints,
  colors,
  components,
  ...themeConfig,
}

**Comment by kristian240** [comment]
> So I can get the error coming through on my IDE which was not happening before:
> 
> ```
> import { extendTheme, ThemeOverride } from "@chakra-ui/react"
> const appTheme: ThemeOverride = {
>   styles,
>   breakpoints,
>   colors,
>   components,
>   ...themeConfig,
> }
> ```

If it helps, I had that issue from the beginning. Maybe the TS version has effect too?

**Comment by TimKolberger** [comment]
I just pushed a fix for this. If you are curious and would like to test this solution, you could replace the type `DeepThemeExtension` in `node_modules/@chakra-ui/react/dist/types/extend-theme.d.ts` with

```ts
declare type DeepThemeExtension<BaseTheme, ThemeType> = {
  [Key in keyof BaseTheme]?: BaseTheme[Key] extends (...args: any[]) => any
    ? DeepThemeExtension<
        Partial<ReturnType<BaseTheme[Key]>>,
        CloneKey<ThemeType, Key>
      >
    : BaseTheme[Key] extends Array<any>
    ? CloneKey<ThemeType, Key>
    : BaseTheme[Key] extends object
    ? DeepThemeExtension<Partial<BaseTheme[Key]>, CloneKey<ThemeType, Key>>
    : CloneKey<ThemeType, Key>;
};
```

**Comment by leggettc18** [comment]
This worked for me! Is there a way to get this fix in a more "permanent" fashion or do I just have to wait for the next release?

**Comment by dodas** [comment]
I'd like to point out to the [`patch-package`](https://www.npmjs.com/package/patch-package) that could help in all cases where you can't wait for some package to ship some fix!
